virttest/utils_iothread.py

```diff
@@ -0,0 +1,211 @@
+        """
+        raise NotImplementedError
+
+    def activate_iothread(self, iothread):
+        """Push iothread to manager."""
+        iothread_id = iothread.get_aid()
+        if self.find_iothread(iothread_id) is not None:
+            raise KeyError("iothread %s already exists" % iothread_id)
+        self.iothread_finder[iothread_id] = iothread
+
+    def deactivate_iothread(self, iothread):
+        """Pop iothread out of manager."""
+        iothread_id = iothread.get_aid()
+        try:
+            self.iothread_finder.pop(iothread_id)
+        except KeyError:
+            raise KeyError("iothread %s not exists" % iothread_id)
+
+
+class PredefinedManager(IOThreadManagerBase):
+    """
+    Support legacy usage of iothread.
+    ```
+    iothreads = iothread0 iothread1
+    iothread_image0 = iothread1
+    ```
+    """
+
+    def request_iothread(self, iothread):
+        """Request predefined iothread."""
+        if iothread == "auto":
+            raise ValueError("Not support 'auto' request mode.")
+        iothread_obj = self.find_iothread(iothread)
+        if iothread_obj:
+            return iothread_obj
+        else:
+            raise KeyError("Unable to find iothread %s" % iothread)
+
+
+class RoundRobinManager(IOThreadManagerBase):
+    """Dispatch iothread object in round-robin way."""
+
+    def __init__(self, iothreads):
+        super(RoundRobinManager, self).__init__(iothreads)
+        self.iothreads = PriorityQueue(
+            iothreads, _iothread_key_getter, _iothread_key_getter)
+
+    def request_iothread(self, iothread):
+        """Return iothread with least device attached."""
+        if iothread == "auto":
+            try:
+                self.iothreads.get_lowest_item()
+            except KeyError:
+                raise KeyError("No available iothread to allocate")
+        else:
+            raise ValueError("Not support request specific iothread")
+
+    def activate_iothread(self, iothread):
+        """Push iothread to manager."""
+        super(RoundRobinManager, self).activate_iothread(iothread)
+        self.iothreads.add_item(iothread)
+
+    def deactivate_iothread(self, iothread):
+        """Pop iothread out of manager."""
+        super(RoundRobinManager, self).deactivate_iothread(iothread)
+        self.iothreads.remove_item(iothread)
+
+
+class OTOManager(IOThreadManagerBase):
+    """Always dispatch new iothread object."""
+
+    def request_iothread(self, iothread):
+        """Return new iothread object."""
+        if iothread == "auto":
+            return self.create_iothread()
+        else:
+            raise ValueError("Not support request specific iothread")
+
+
+class OptimalManager(RoundRobinManager):
+    """Dispatch iothread based on min(vcpu, pcpu, device)"""
+
+    def __init__(self, vcpu):
+        super(OptimalManager, self).__init__([])
+        self.cpu_bound = min(vcpu, cpu.total_cpus_count())
```

Comment:
In case of determining the number of iothread across multiple hosts (i.e. migration scenarios), it cannot only reference by the resources of localhost, hence, `pcpu` should be given by users.